File: lib/services/folder_service.dart
import 'package:uuid/uuid.dart';
import '../database/database.dart';
import '../database/daos.dart';
import 'note_service.dart';
import 'sync_op_recorder.dart';

class FolderService {
  final FolderDao folderDao;
  final NoteService noteService;
  final SyncOpRecorder _recorder;
  final _uuid = const Uuid();

  FolderService(this.folderDao, this.noteService, this._recorder);

  Stream<List<FolderWithCount>> watchFolders(String userId) {
    return folderDao.watchFoldersWithNoteCount(userId);
  }

  Future<int> createFolder(String name, String userId) async {
    final id = _uuid.v4();
    final folder = Folder(id: id, name: name, userId: userId);
    final rowId = await folderDao.insertFolder(
      FoldersCompanion.insert(
        id: id,
        name: name,
        userId: userId,
      ),
    );
    await _recorder.folderCreated(folder);
    return rowId;
  }

  Future renameFolder(Folder folder, String newName) async {
    final updated = folder.copyWith(name: newName);
    await folderDao.updateFolder(updated);
    await _recorder.folderRenamed(updated);
  }

  Future deleteFolder(Folder folder) async {
    // Move child notes back to root first, otherwise they'd be invisible
    await noteService.clearFolderFromNotes(folder.id);
    await folderDao.deleteFolder(folder);
    await _recorder.folderHardDeleted(folder);
  }
}

File: note-desktop/Sources/Services/FolderService.swift
import Foundation

public final class FolderService {
    private let storage: StorageService
    private let noteService: NoteService
    private let recorder: SyncOpRecorder
    
    public init(storage: StorageService, noteService: NoteService, recorder: SyncOpRecorder) {
        self.storage = storage
        self.noteService = noteService
        self.recorder = recorder
    }
    
    public func createFolder(name: String, userId: String) -> DBFolder? {
        let folderId = UUID().uuidString
        let folder = DBFolder(id: folderId, name: name, userId: userId)
        
        guard storage.insertFolder(folder) else { return nil }
        _ = recorder.folderCreated(folder)
        return folder
    }
    
    public func renameFolder(_ folder: DBFolder, newName: String) -> Bool {
        var updated = folder
        updated.name = newName
        guard storage.updateFolder(updated) else { return false }
        _ = recorder.folderRenamed(updated)
        return true
    }
    
    public func deleteFolder(_ folder: DBFolder) -> Bool {
        // Move child notes back to root first
        _ = storage.clearFolderFromNotes(folderId: folder.id)
        guard storage.deleteFolder(id: folder.id) else { return false }
        _ = recorder.folderHardDeleted(folder)
        return true
    }
}
